### PR TITLE
Implement stand-alone open/next/close iterators in C++

### PR DESCRIPTION
**Pull request by ingomueller-net**
This PR adds a very simple stand-alone project in `experimental` that illustrates how database/relational operators (aka open/next/close iterators or "volcano" iterators) could be implemented if each query was written in C++. The code is quite work-in-progress: there are only to different iterators, the build process is not integrated with that of the remainder of the repository, and many other things I haven't thought about are missing; however, they should allow us to point to some of the concepts of iterators for implementing them in MLIR.

### CODE REVIEW

**Comment by ingomueller-net** [comment]
@nicolasvasilache: This is ready for a first look.

File: experimental/iterators/.clang-format [hunk 1]
```diff
@@ -0,0 +1,2 @@
+---
+BasedOnStyle: Google
```
Comment (nicolasvasilache) [inline experimental/iterators/.clang-format:2]: Note, we generally follow LLVM's coding style: https://llvm.org/docs/CodingStandards.html

While things are under experimental, no need to change.

File: experimental/iterators/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,43 @@
+cmake_minimum_required(VERSION 3.0.0)
+project(iterators VERSION 0.1.0)
+
+set(CMAKE_CXX_STANDARD 17)
+
+# Set up GoogleTest dependency
+include(FetchContent)
+FetchContent_Declare(
```
Comment (nicolasvasilache) [inline experimental/iterators/CMakeLists.txt:8]: Traditionally I'd have gone for a git submodule but this seems simple and fine for experimental.

One thing to note is that in the future we will want to impl. tests using FileCheck (https://llvm.org/docs/CommandGuide/FileCheck.html). 

See for example this somewhat complex test (https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/benchmark_matmul.mlir#L98) which starts from MLIR, calls into C / C++ to verify and print 0 on success, then performs checks with FileCheck (https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/benchmark_matmul.mlir#L91).

No need to change anything here, just mentioning so you can think ahead in terms compatible with the current best practices.

File: experimental/iterators/include/operators/column_scan.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  int64_t current_pos_;
+};
+
+template <typename... InputTypes>
+auto MakeColumnScanOperator(std::vector<InputTypes>... inputs) {
+  return ColumnScanOperator<InputTypes...>(inputs...);
+}
+
+
+#endif  // OPERATORS_COLUMN_SCAN_H
```
Comment (nicolasvasilache) [inline experimental/iterators/include/operators/column_scan.h:50]: newline here andeverywhere

File: experimental/iterators/include/operators/column_scan.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      const std::index_sequence<kIndices...> & /*unused*/) const {
+    return std::make_tuple(std::get<kIndices>(inputs_).at(current_pos_)...);
+  }
+
+  std::tuple<std::vector<InputTypes>...> inputs_;
+  int64_t current_pos_;
+};
+
+template <typename... InputTypes>
+auto MakeColumnScanOperator(std::vector<InputTypes>... inputs) {
```
Comment (nicolasvasilache) [inline experimental/iterators/include/operators/column_scan.h:45]: Could you add some comments to explain the expected model of computation?

I am inferring that you want a "struct of array" input (or a tuple<vector<Type>>) and you are going to coiterate on all these and extract a tuple<Type>.

Using array of struct / struct of array terminology (in comments, in addition to DB terminology) would be nice to help brige the naming gap.

File: experimental/iterators/include/operators/column_scan.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+class ColumnScanOperator {
+ public:
+  using OutputTuple = std::tuple<InputTypes...>;
+  using ReturnType = std::optional<OutputTuple>;
+
+  explicit ColumnScanOperator(std::vector<InputTypes>... inputs)
+      : inputs_(std::make_tuple(std::move(inputs)...)), current_pos_(0) {}
+
+  void Open() {}
+  ReturnType ComputeNext() {
```
Comment (nicolasvasilache) [inline experimental/iterators/include/operators/column_scan.h:18]: one (the main) implication of the LLVM style here would be that the functions and variables are all snakeCase.

**Review by nicolasvasilache — APPROVED** [review]
Thanks!

Generally looks good, comments/doc would be welcome to explain e.g. how data is expected to be formatted in the data source, how it can be retrieved (e.g. one at a time, in chunks) and how we expect it to be processed. Not everything needs to happen in this first PR though!

I'll ping you this pm to chat a bit and see where to go from here (unless you already started on next steps).

File: experimental/iterators/.clang-format (hunk 1 above)
Comment (ingomueller-net) [reply experimental/iterators/.clang-format:2]: OK, sure, if the code lives here, then the LLVM style makes sense.

File: experimental/iterators/CMakeLists.txt (hunk 2 above)
Comment (ingomueller-net) [reply experimental/iterators/CMakeLists.txt:8]: I had skimmed the repo for tests and found the rewrite tests I think you mention. However, what I am testing currently doesn't seem to apply. It sure does once I get to the MLIR dialect, so it's good to have the pointers. Maybe we can discuss options later online.

File: experimental/iterators/include/operators/column_scan.h (hunk 3 above)
Comment (ingomueller-net) [reply experimental/iterators/include/operators/column_scan.h:50]: Oops, didn't see this one. Which other new lines are problematic?

File: experimental/iterators/CMakeLists.txt (hunk 2 above)
Comment (nicolasvasilache) [reply experimental/iterators/CMakeLists.txt:8]: sg!

File: experimental/iterators/.clang-format (hunk 1 above)
Comment (ingomueller-net) [reply experimental/iterators/.clang-format]: The earlier fix this, the cheaper --> changed with the latest commit.

File: experimental/iterators/include/operators/column_scan.h (hunk 3 above)
Comment (nicolasvasilache) [reply experimental/iterators/include/operators/column_scan.h]: I meant trailing newline, if you look at the github diff, it shows a 🚫 sign where you are missing a newline at the end of the file

File: experimental/iterators/include/operators/column_scan.h (hunk 5 above)
Comment (ingomueller-net) [reply experimental/iterators/include/operators/column_scan.h]: Dito: cheapest to fix now. The latest commit applies all things I realized deviated from the LLVM style guide. Let me know if you spot anything else (or I misunderstood the guide or whatever).

File: experimental/iterators/include/operators/column_scan.h (hunk 3 above)
Comment (ingomueller-net) [reply experimental/iterators/include/operators/column_scan.h]: Got it! Fixed.

File: experimental/iterators/include/operators/column_scan.h (hunk 4 above)
Comment (ingomueller-net) [reply experimental/iterators/include/operators/column_scan.h]: I have added documentation with the latest commit. Let me know if you spot anything I should add.

File: experimental/iterators/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+stream has been reached, i.e., no further tuples can be returned.
+
+Each operator has a `Make*Operator` factory function that derives the template
+parameters for the to-be-instatiated class such that assembling query plans is
+concise:
+
+```cpp
+std::vector<int32_t> Numbers = {1, 2, 3, 4};
+auto Scan = MakeColumnScanOperator(Numbers);
+auto Reduce = MakeReduceOperator(&Scan, [](auto T1, auto T2) {
```
Comment (nicolasvasilache) [inline experimental/iterators/README.md:54]: nit: MLIR coding standards require snakeCase for variables and function names.
so: `numbers`, `scan`, `makeColumnScanOperator` etc

File: experimental/iterators/include/operators/column_scan.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    CurrentPos++;
+    return Ret;
+  }
+
+  /// Does nothing.
+  void close() {}
+
+private:
+  template <std::size_t... kIndices>
+  std::optional<OutputTuple> ExtractCurrentTuple(
```
Comment (nicolasvasilache) [inline experimental/iterators/include/operators/column_scan.h:57]: nit: this should also be snakeCase

File: experimental/iterators/include/operators/column_scan.h [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  std::tuple<std::vector<InputTypes>...> Inputs;
+  int64_t CurrentPos;
+};
+
+/// Creates a new `ColumnScanOperator` deriving its template parameters from
+/// the provided arguments.
+template <typename... InputTypes>
+auto MakeColumnScanOperator(std::vector<InputTypes>... Inputs) {
```
Comment (nicolasvasilache) [inline experimental/iterators/include/operators/column_scan.h:69]: nit: this too

**Review by nicolasvasilache — APPROVED** [review]
LGTM but please fix formatting towards snakeCase

File: experimental/iterators/.clang-format (hunk 1 above)
Comment (nicolasvasilache) [reply experimental/iterators/.clang-format]: snakeCasing is only applied in a few places at this point though

**Pull request merged**